core/utils/cuckoo_map.h

```diff
@@ -0,0 +1,331 @@
+      }
+    }
+    return -1;
+  }
+
+  int MakeSpace(uint32_t index, int depth) {
+    if (depth >= kMaxCuckooPath) {
+      return -1;
+    }
+
+    Bucket& bucket = buckets_[index];
+
+    for (int i = 0; i < kEntriesPerBucket; i++) {
+      const K& key = bucket.key_indices[i];
+      uint32_t pri = HashNonZero(key);
+      uint32_t sec = HashSecondary(pri);
+
+      uint32_t alt_index;
+
+      // this entry is in its primary bucket?
+      if (pri == bucket.hash_values[i]) {
+        alt_index = sec & bucket_mask_;
+      } else if (sec == bucket.hash_values[i]) {
+        alt_index = pri & bucket_mask_;
+      } else {
+        return -1;
+      }
+
+      // Find empty slot
+      int j = FindEmptySlot(alt_index);
+      if (j == -1) {
+        j = MakeSpace(alt_index, depth + 1);
+      }
+      if (j >= 0) {
+        Bucket& alt_bucket = buckets_[alt_index];
+        alt_bucket.hash_values[j] = bucket.hash_values[i];
+        alt_bucket.key_indices[j] = bucket.key_indices[i];
+        bucket.hash_values[i] = 0;
+        return i;
+      }
+    }
+
+    return -1;
+  }
+
+  Entry* GetHash(uint32_t primary, const K& key) {
+    Entry* ret = GetFromBucket(primary, primary, key);
+    if (ret) {
+      return ret;
+    }
+    return GetFromBucket(primary, HashSecondary(primary), key);
+  }
+
+  Entry* GetFromBucket(uint32_t primary, uint32_t hash_val, const K& key) {
+    Bucket& bucket = buckets_[hash_val & bucket_mask_];
+    for (int i = 0; i < kEntriesPerBucket; i++) {
+      if (primary != bucket.hash_values[i]) {
+        continue;
+      }
+      uint32_t idx = bucket.key_indices[i];
+      if (C(entries_[idx].first, key)) {
+        return &entries_[idx];
+      }
+    }
+    return nullptr;
+  }
+
+  static uint32_t MakeNonZero(uint32_t v) {
+    // Set the MSB and unset the 2nd MSB (NOTE: must be idempotent).
+    // Then the result will never be a zero, and not a non-number when
+    // represented as float (so we are good to use _mm_*_ps() SIMD ops)
+    return (v | (1u << 31)) & (~(1u << 30));
+  }
+
+  static uint32_t HashSecondary(uint32_t primary) {
+    uint32_t tag = primary >> 12;
+    return primary ^ ((tag + 1) * 0x5bd1e995);
+  }
+
+  static uint32_t Hash(const K& key) { return H(key, kHashInitval); }
+
+  static uint32_t HashNonZero(const K& key) { return MakeNonZero(Hash(key)); }
+
+  void ExpandEntries() {
+    uint32_t old_size = num_entries_;
+    uint32_t new_size = old_size + old_size / 2;
+
+    entries_ = reinterpret_cast<Entry*>(
+        mem_realloc(entries_, new_size * sizeof(Entry)));
+    num_entries_ = new_size;
+
+    for (uint32_t i = new_size - 1; i >= old_size; --i) {
+      free_entry_indices_.push(i);
+    }
+  }
+
+  void ExpandBuckets() {
+    uint32_t new_size = (bucket_mask_ + 1) * 2;
+    bucket_mask_ = new_size - 1;
+
+    buckets_ = reinterpret_cast<Bucket*>(
+        mem_realloc(buckets_, new_size * sizeof(Bucket)));
+  }
+
+  // Tunable macros
+  static const int kInitNumBucket = 4;
```

Comment:
types and constants should appear before member functions (if i remember correctly)